resolver/blocking_resolver.go

```diff
@@ -385,8 +385,14 @@ func (r *BlockingResolver) handleDenylist(ctx context.Context, groupsToCheck []s
 			return true, resp, err
 		}
 
-		if groups := r.matches(groupsToCheck, r.denylistMatcher, domain); len(groups) > 0 {
-			resp, err := r.handleBlocked(logger, request, question, fmt.Sprintf("BLOCKED (%s)", strings.Join(groups, ",")))
+		if matches := r.matches(groupsToCheck, r.denylistMatcher, domain); len(matches) > 0 {
+			var entries []string
+			for group, rule := range matches {
+				entries = append(entries, strings.Join([]string{group, rule}, ":"))
+			}
+
+			msg := fmt.Sprintf("BLOCKED (%s)", strings.Join(entries, ", "))
+			resp, err := r.handleBlocked(logger, request, question, msg)
```

Comment:
Can you pass `matches` to `handleBlock` instead of `msg` so we can move the formatting there and deduplicate it with `Resolve`?

Reply:
resolver/blocking_resolver.go:383 passes a string:
```go
resp, err := r.handleBlocked(logger, request, question, "BLOCKED (ALLOWLIST ONLY)")
```

Could we return custom type from `r.matches(...)` and override `String()` method?